freva-rest/src/freva_rest/config.py

```diff
@@ -20,13 +20,36 @@
 
 CONFIG_TYPE = TypedDict(
     "CONFIG_TYPE",
-    {"API_CONFIG": Path, "LOGGER": logging.Logger, "NAME": str, "DEBUG": bool},
+    {
+        "API_CONFIG": Path,
+        "LOGGER": logging.Logger,
+        "NAME": str,
+        "DEBUG": bool,
+        "API_CACHE_EXP": int,
+        "API_URL": str,
+        "REDIS_HOST": Optional[str],
+        "REDIS_SSL_CERTFILE": Optional[str],
+        "REDIS_SSL_KEYFILE": Optional[str],
+        "REDIS_PASS": Optional[str],
+        "REDIS_USER": Optional[str],
+        "KEYCLOAK_HOST": str,
+        "KEYCLOAK_REALM": str,
```

Comment:
What are your thoughts to consider `KEYCLOAK_REALM` optional? because from my point of experience in Helmholtz, they were not using any realm arg for their opneid configurations urls:

https://login-dev.helmholtz.de/oauth2/.well-known/openid-configuration

https://login.helmholtz.de/oauth2/.well-known/openid-configuration

with this idea, users with these conditions, can get benefit of authentication of Freva.
my idea:

```python
CONFIG_TYPE = TypedDict(
    "CONFIG_TYPE",
    {
        "API_CONFIG": Path,
        "LOGGER": logging.Logger,
        "NAME": str,
        "DEBUG": bool,
        "API_CACHE_EXP": int,
        "API_URL": str,
        "REDIS_HOST": Optional[str],
        "REDIS_SSL_CERTFILE": Optional[str],
        "REDIS_SSL_KEYFILE": Optional[str],
        "REDIS_PASS": Optional[str],
        "REDIS_USER": Optional[str],
        "KEYCLOAK_HOST": str,
        "KEYCLOAK_REALM": Optional[str],
        "KEYCLOAK_CONF": Optional[str],
    },
)
defaults: CONFIG_TYPE = {
    "API_CONFIG": Path(__file__).parent / "api_config.toml",
    "LOGGER": logger,
    "NAME": THIS_NAME,
    "DEBUG": False,
    "API_CACHE_EXP": 3600,
    "REDIS_SSL_CERTFILE": None,
    "REDIS_SSL_KEYFILE": None,
    "REDIS_HOST": None,
    "REDIS_USER": None,
    "REDIS_PASS": None,
    "KEYCLOAK_HOST": "http://localhost:8080",
    "KEYCLOAK_REALM": "freva",
    "API_URL": "http://localhost:7777",
    "KEYCLOAK_CONF": None,
}

    @staticmethod
    def get_keycloak_url() -> str:
        """Construct the keycloak realm url."""
        keycloak_host = os.getenv("KEYCLOAK_HOST", defaults["KEYCLOAK_HOST"])
        keycloak_realm = os.getenv("KEYCLOAK_REALM", defaults["KEYCLOAK_REALM"])
        keycloak_conf = os.getenv("KEYCLOAK_CONF", defaults["KEYCLOAK_CONF"])
        _, split, host = keycloak_host.partition("://")
        if keycloak_conf:
            return f"{keycloak_conf.replace("/.well-known/openid-configuration", "")}"
        if split:
            host = keycloak_host
        else:
            host = f"https://{keycloak_host}"
        return f"{host}/realms/{keycloak_realm}"
```

Reply:
But this is server side, things are less configurable. And I don't want to make the realm configurable for the user. Because things need to be set up on keycloak for that. But maybe I don't understand entirely.

Reply:
Alright, since this is important, let me explain a bit further.
The main goal here is to ensure that the authentication for Freva can comply with the newest strict security implemented here, and on the other hand also making the entire Freva instance as configurable and flexible as possible with different OpenID instances.

We need to clarify here that the users in this context are not the individuals running plugins or using Freva directly. Instead, the users are coordinators and administrators who manage to deploy Freva instance for their institution or projects. Via my proposed idea each institution that is running Freva has the option to configure their realm or not, depends how they defined their authentication and how they redirected their Nginx, which in result shows Freva's flexibility to work in different OpenID scenarios and situations.

For example, as I mentored in my comment, an institute is part of Helmholtz and they want run Freva instance and use the [Helmhotlz AAI](https://hifis.net/aai/) (OpenID of whole institutions in HIFS such as Hereon). In this case, as I mentioned their endpoints, they wouldn't define a realm in the configuration, and with this solution allowing them for adjustments to be fit with Freva. Additionally, there might be other setups where other institutions prefer to redirect their keycloak endpoints via NGINX configuration to some other links that I don't know at the moment. In such cases, this flexibility would be beneficial for FrevaAPI.

Due to importance of this matter, I think it needs further discussion ...

Reply:
And as another idea for future development, we can take advantage of [organizations](https://www.keycloak.org/2024/06/announcement-keycloak-organizations) in keycloak to restrict and manage the users for getting them access to the specific instance of Freva ...

Reply:
Another point that just crossed to my mind, what r ur thought, FrevaAPI supports different OpneID instacnes in one Freva instacne? For example a Freva instance can be able to login via a list of `return [f"{host}/realms/{keycloak_realm}" ]` well-configs and then as result, based on some criteria it looks for the token endpoint and other needed endpoints there? it means, for example, one Freva instacne, can support DWD, DKRZ and other keycloaks in a same time!

Reply:
Yes, that is a very good point you are raising. We had a meeting with DWD and FUB on keycloak. The FUB admin raised the point that it would be great to support different SSO instances. That's something we need to keep in mind. But I am not sure if we should implement it straight away or aim at getting this done later.

Reply:
i think additionally we need a loop over `oauth2_scheme` as well ...

Reply:
Or as another idea, we can consider the `client_id` and `client_pass` as tuples, then user with the `username` and `password`, looks for it's proper SSO automatically, among a tuple of SSOs! it seems more realistic ...

Reply:
The organisation idea is a good one, We could indeed try to implement that. I don't think we need multiple realms as DWD or FUB will never ever outsource the authentication mechanism to us. But we keep the orga idea for multiple instances in mind. 

I'd be inclined to say we merge this first and do this afterwards.